libsqlite3-sys/src/lib.rs

```diff
@@ -71,3 +71,64 @@ impl Default for sqlite3_vtab_cursor {
         unsafe { mem::zeroed() }
     }
 }
+
+#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+mod allocator {
```

Comment:
This allocator seems broken. sqlite_free assumes that the first 8 bytes of the pointer are going to be the size, but sqlite_malloc doesn't allocate space for that, nor store it, etc. I'm surprised this works at all (it seems like if it does, it's a coincidence based on an implementation detail of SQLite allocations).

You need something more along the lines of https://shift.click/blog/on-dealloc/#appendix-how-to-wrap-allocators-which-need-layout-info-in-dealloc, where the information about the allocation is stored before the pointer. Note that that code also handles the full layout, e.g. alignment too, whereas for this we can assume the needed alignment is always the same, and only handle the size.

Reply:
https://sqlite.org/malloc.html#the_default_memory_allocator :

>The default allocator implements memsize() by always allocating 8 extra bytes on each malloc() request and storing the size of the allocation in that 8-byte header.